arrow/arrowhead.pyx

```diff
@@ -0,0 +1,149 @@
+            double[::1] rates not None,
+
+            int64_t[::1] reactants_lengths not None,
+            int64_t[::1] reactants_indexes not None,
+            int64_t[::1] reactants not None,
+            int64_t[::1] reactions not None,
+
+            int64_t[::1] dependencies_lengths not None,
+            int64_t[::1] dependencies_indexes not None,
+            int64_t[::1] dependencies not None,
+
+            int64_t[::1] substrates_lengths not None,
+            int64_t[::1] substrates_indexes not None,
+            int64_t[::1] substrates not None,
+            *args):
+        self.random_seed = random_seed
+        mersenne.seed(self.info.random_state, random_seed)
+
+        self.info.reactions_count = stoichiometry.shape[0]
+        self.info.substrates_count = stoichiometry.shape[1]
+        self.info.stoichiometry = &stoichiometry[0, 0]
+        self.info.rates = &rates[0]
+
+        self.info.reactants_lengths = &reactants_lengths[0]
+        self.info.reactants_indexes = &reactants_indexes[0]
+        self.info.reactants = &reactants[0]
+        self.info.reactions = &reactions[0]
+
+        self.info.dependencies_lengths = &dependencies_lengths[0]
+        self.info.dependencies_indexes = &dependencies_indexes[0]
+        self.info.dependencies = &dependencies[0]
+
+        self.info.substrates_lengths = &substrates_lengths[0]
+        self.info.substrates_indexes = &substrates_indexes[0]
+        self.info.substrates = &substrates[0]
+
+        # TODO(jerry): Check array sizes? E.g. rates.shape[0] >= reactions_count
+
+        self.refs = (  # hold refs to these memoryviews while we hold their data ptrs
+            stoichiometry, rates,
+            reactants_lengths, reactants_indexes, reactants, reactions,
+            dependencies_lengths, dependencies_indexes, dependencies,
+            substrates_lengths, substrates_indexes, substrates)
+
+    # `self` might no longer be a valid Python object so *BE CAREFUL* invoking
+    # Python operations that might touch it!
+    def __dealloc__(self):
+        PyMem_Free(self.info.random_state)
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    def evolve(self, double duration, int64_t[::1] state):
+        """Run the Gillespie algorithm with the initialized stoichiometry and
+        rates over the given duration and state.
+        Return None or a tuple (steps, time, events, outcome).
+        """
+        # TODO(jerry): Check the state[] array size?
+
+        evolved = obsidian.evolve(&self.info, duration, &state[0])
+        cdef int steps = evolved.steps
+        cdef int count = self.info.substrates_count
+
+        if steps == -1:
+            return None
+
+        time = copy_c_array(evolved.time, steps, sizeof(double), np.NPY_DOUBLE)
+        events = copy_c_array(evolved.events, steps, sizeof(int64_t), np.NPY_INT64)
+        outcome = copy_c_array(evolved.outcome, count, sizeof(int64_t), np.NPY_INT64)
```

Comment:
It still seems slightly clumsy we have to copy the results into new arrays and then free the old arrays, but I can't think of a better way to do it right now.

Reply:
Agreed. `obsidian.c` is defined with a straight C API, so it uses `malloc` to allocate return arrays then hands off ownership to the caller.

So here are the options I considered:

* Keep the existing, reasonable approach: Copy the array contents into numpy arrays. (It could equally be `array.array` values or lists or other Python objects, but that'd change the API for no good reason.)
* Pass in an array allocator function for `obsidian.c` to call. This has the potential to improve performance, esp. if it can keep reusing the same arrays and occasionally grow them.
* If the caller could predict the needed array sizes, or conservatively make an acceptable overestimate, it could allocate the arrays and pass them in for `obsidian.c` to fill in. This seems like a bunch of work for unlikely payoff.
* In numpy 1.17 it's possible to construct an array around a given buffer and hand off buffer ownership, providing a deallocator callback.

BTW `Arrowhead` uses the Python memory allocator `PyMem_Malloc()`/`PyMem_Free()` where possible. It's tuned for Python usage, e.g. lots of small nodes, and I think it tracks total Python memory use. But it has to use the C `free()` for the nodes allocated by `obsidian.c`.